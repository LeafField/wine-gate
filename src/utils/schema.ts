import { z } from "zod";

export const searchSchema = z.object({
  search: z
    .string()
    .min(4, { message: "4文字以上入力してください" })
    .max(60, { message: "60文字以下で入力してください" }),
  category: z.string(),
  sort: z.enum(["new", "inexpensive", "popular"]),
});
