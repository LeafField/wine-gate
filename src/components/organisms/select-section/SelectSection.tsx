import React, { FC, PropsWithChildren } from "react";
import SmallArticle from "../../molecules/small-article/SmallArticle";
import MoreLink from "../../atoms/more-link/MoreLink";
import { CategoryFavoriteSmallArticleProps } from "../../../types/article_types";

type Props = {
  wines: CategoryFavoriteSmallArticleProps;
  title: string;
  link: string;
};

const SelectSection: FC<PropsWithChildren<Props>> = ({
  wines,
  children,
  title,
  link,
}) => {
  return (
    <section className="mt-6 max-w-[1392px] border-b border-gray pb-6">
      <h4 className="text-heading3">{title}</h4>
      <p className="text-main-text">{children}</p>
      <div className="mt-6 grid grid-cols-1 gap-x-5 gap-y-4 tablet:grid-cols-[repeat(auto-fill,333px)]">
        {wines.map((content) => (
          <SmallArticle
            key={content.id}
            id={content.id}
            categories={{ chara: content.categories_chara }}
            image_src={content.image_src}
            tags={content.tags}
            title={content.title}
          />
        ))}
      </div>
      <MoreLink link={link} />
    </section>
  );
};

export default SelectSection;
