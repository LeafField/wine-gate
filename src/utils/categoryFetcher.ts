import { supabase } from "./supabase";
import { cache } from "react";
import { ArticleProps } from "../types/article_types";

export const categoryFetcher = cache(
  async (slug: string[]): Promise<ArticleProps[] | null> => {
    if (slug.length !== 3) return null;
    const category = slug[0];
    const sort = slug[1];
    const page = Number(slug[2]);
    const start = (page - 1) * 10;
    const end = page * 10 - 1;
    if (sort === "new") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("created_at", { ascending: false })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "inexpensive") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("price", { ascending: true })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "popular") {
      const { data, error } = await supabase
        .rpc("get_wine_category_favorite", {
          category_value: category,
        })
        .limit(10);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      const articles: ArticleProps[] = data.map((article) => ({
        categories: {
          category: article.category_name,
          id: article.category_id,
          chara: article.category_chara,
          sub: article.category_sub,
        },
        author_name: article.author_name,
        fruity: article.fruity,
        id: article.id,
        image_src: article.image_src,
        price: article.price,
        sober_or_sweet: article.sober_or_sweet,
        tags: article.tags,
        tart: article.tart,
        title: article.title,
      }));

      return articles;
    }
    return null;
  },
);

export const categoryCountFetcher = cache(async (category: string) => {
  const { count } = await supabase
    .from("wines")
    .select("id,categories(category)", { count: "exact", head: true })
    .eq("categories.category", category);
  if (count) {
    const totalPage = Math.ceil(count / 10);
    return totalPage;
  } else {
    return 1;
  }
});
