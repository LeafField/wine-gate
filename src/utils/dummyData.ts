import dummyImage from "../images/dummy_wine.png";
import {
  SmallArticleProps,
  CategoryProps,
  WineProps,
  ArticleProps,
} from "../types/article_types";

export const smallArticleDummyData: SmallArticleProps[] = [
  {
    id: "1",
    title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
    chara: "渋い",
    tags: "樽熟成",
    image_src: dummyImage.src,
  },
  {
    id: "2",
    title: "コノスル シャルドネ ビシクレタ・レゼルバ",
    chara: "辛口白",
    tags: "フルーティ",
    image_src: dummyImage.src,
  },
  {
    id: "3",
    title: "コノスル カルメネール ビシクレタ・レゼルバ",
    chara: "酸っぱい",
    tags: "樽熟成,濃厚",
    image_src: dummyImage.src,
  },
];

export const categoryDummyData: CategoryProps[] = [
  {
    id: 1,
    category: "sober",
    chara: "渋い",
    sub: "red",
  },
  {
    id: 2,
    category: "tart",
    chara: "酸っぱい",
    sub: "red",
  },
  {
    id: 3,
    category: "fruity",
    chara: "果実味豊か",
    sub: "red",
  },
  {
    id: 4,
    category: "white",
    chara: "辛口白",
    sub: "white",
  },
  {
    id: 5,
    category: "sweet",
    chara: "甘口白",
    sub: "white",
  },
  {
    id: 6,
    category: "rose",
    chara: "ロゼ",
    sub: "interim",
  },
  {
    id: 7,
    category: "orange",
    chara: "オレンジ",
    sub: "interim",
  },
  {
    id: 8,
    category: "sparkling",
    chara: "スパークリング",
    sub: "other",
  },
  {
    id: 9,
    category: "fortified",
    chara: "酒精強化",
    sub: "other",
  },
];

export const selectDummyData = categoryDummyData.map((category) => {
  return { value: category.category, label: category.chara };
});

export const wineDummyData: WineProps = {
  id: "sdfaf",
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_id: "fsdaf",
  author_name: "ワイン大好き太郎",
  image_src: dummyImage.src,
  price: 2000,
  categories: categoryDummyData[0],
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  description:
    "チリのワインといえばコノスル！\nカベルネソーヴィニヨンの中でも特におすすめの一本です。",
  place: "スーパー,ネット通販",
  erudition:
    "カベルネソーヴィニヨンはボルドー地方の赤ワインの原料としても有名です。",
  category_id: 1,
  created_at: "2021-10-10",
};

export const articleDummyData: ArticleProps = {
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_name: "ワイン大好き太郎",
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  categories: categoryDummyData[0],
  id: "1",
  price: 1500,
};
