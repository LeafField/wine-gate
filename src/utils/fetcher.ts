import { supabase } from "./supabase";
import { EditingPageSchemaType, ImageInputSchemaType } from "./schema";
import { User } from "../store";
import { revalidateServer } from "../server/revalidate";

export const postWine = async ({
  value,
  user,
  image,
}: {
  value: EditingPageSchemaType;
  user: User;
  image: ImageInputSchemaType;
}) => {
  if (!image) throw new Error("画像が必須です");
  const IMAGE_SRC = `${user.id}/${image.name}`;

  const { error: imageError, data: imageData } = await supabase.storage
    .from("wines")
    .upload(IMAGE_SRC, image);

  if (imageError) throw new Error("画像のアップロードに失敗しました");
  if (imageData) {
    const { data, error } = await supabase
      .from("wines")
      .insert({
        title: value.title.replace(/　/g, " "),
        price: value.price,
        place: value.place,
        description: value.description,
        erudition: value.erudition,
        category_id: Number(value.category_id),
        fruity: value.fruity,
        tart: value.tart,
        sober_or_sweet: value.sober_or_sweet,
        tags: value.tags.join(),
        author_id: user.id,
        author_name: user.username,
        image_src: imageData.path,
      })
      .select();
    if (error) throw new Error("投稿に失敗しました");
    return data[0].id;
  }
};

export const updateWine = async ({
  id,
  image,
  oldWineImage,
  user_id,
  value,
}: {
  id: string;
  user_id: string;
  value: EditingPageSchemaType;
  image: ImageInputSchemaType | null;
  oldWineImage: string;
}) => {
  revalidateServer(id);
  if (image) {
    const IMAGE_SRC = `${user_id}/${image.name}`;
    const { error } = await supabase.storage
      .from("wines")
      .upload(IMAGE_SRC, image);
    if (error) throw new Error("画像のアップロードに失敗しました");
    const { error: deleteError } = await supabase.storage
      .from("wines")
      .remove([oldWineImage]);
    if (deleteError) throw new Error("画像の削除に失敗しました");
  }
  const { data, error } = await supabase
    .from("wines")
    .update({
      title: value.title.replace(/　/g, " "),
      price: value.price,
      place: value.place,
      description: value.description,
      erudition: value.erudition,
      category_id: Number(value.category_id),
      fruity: value.fruity,
      tart: value.tart,
      sober_or_sweet: value.sober_or_sweet,
      tags: value.tags.join(),
      image_src: image ? `${user_id}/${image.name}` : oldWineImage,
    })
    .eq("id", id)
    .select();
  if (error) throw new Error("ワインの更新に失敗しました");
  return data[0].id;
};

export const favoriteCount = async (wine_id: string) => {
  const { count, error } = await supabase
    .from("favorite")
    .select("id", { count: "exact", head: true })
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのカウントに失敗しました");
  return count;
};

export const favoriteUserOne = async (wine_id: string) => {
  const { data, error } = await supabase
    .from("favorite")
    .select("user_id")
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのユーザー取得に失敗しました");
  return data[0] ? data[0] : null;
};

export const addFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .insert({ wine_id, user_id });

  if (error) throw new Error("お気に入りの登録に失敗しました");
};

export const removeFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .delete()
    .eq("wine_id", wine_id)
    .eq("user_id", user_id);

  if (error) throw new Error("お気に入りの解除に失敗しました");
};

export const getWine = async (id: string) => {
  const { data, error } = await supabase
    .from("wines")
    .select("*,categories(*)")
    .eq("id", id)
    .single();
  if (error) throw new Error("ワインの取得に失敗しました");

  return data;
};
