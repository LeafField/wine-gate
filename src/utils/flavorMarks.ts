import { ReactNode } from "react";

type Marks = {
  value: number;
  label: ReactNode;
};

const sweet: Marks[] = [
  {
    value: 1,
    label: "辛口",
  },
  {
    value: 2,
    label: "中辛",
  },
  {
    value: 3,
    label: "中口",
  },
  {
    value: 4,
    label: "中甘",
  },
  {
    value: 5,
    label: "甘口",
  },
];

export const tart: Marks[] = [
  {
    value: 1,
    label: "ない",
  },
  {
    value: 2,
    label: "弱い",
  },
  {
    value: 3,
    label: "普通",
  },
  {
    value: 4,
    label: "強め",
  },
  {
    value: 5,
    label: "酸っぱい",
  },
];
