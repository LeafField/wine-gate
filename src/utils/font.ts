import { Noto_Sans_JP, Kumar_One } from "next/font/google";

export const notoStyle = Noto_Sans_JP({ subsets: ["latin"] });
export const KumarOne = Kumar_One({ subsets: ["latin"], weight: "400" });
