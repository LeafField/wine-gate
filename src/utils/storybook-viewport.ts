export const mobileViewport = {
  viewport: {
    defaultViewport: "iphone6",
  },
};
