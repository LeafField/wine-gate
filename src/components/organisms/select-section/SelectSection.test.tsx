import { screen, render } from "@testing-library/react";
import SelectSection from "./SelectSection";
import { categoryFavoriteSmallArticleDummyData } from "../../../utils/dummyData";

describe("SelectSectionの単体テスト", () => {
  test("与えられたtitle及びchildren要素を元にコンポーネントが正しく描画される", () => {
    const { container } = render(
      <SelectSection
        wines={categoryFavoriteSmallArticleDummyData}
        title="渋い"
        link="sober"
      >
        タンニンが強さが強調されていたり、ブラックベリーやカシス、プルーン等黒系果実で表現される事も多い種類です。
        <br />
        タンニンは口の中で渋みを感じさせる成分であり、若いワインでは特に強く感じられることがあります。
        <br />
        カベルネ・ソーヴィニヨンやシラーなどの品種が一般的にこのタイプのワインに含まれます。
        <br />
        主に色が濃く重めの肉料理に合いやすいです。
      </SelectSection>,
    );
    expect(screen.getByRole("heading", { name: "渋い" })).toBeInTheDocument();
    expect(
      screen.getByText(
        /タンニンが強さが強調されていたり、ブラックベリーやカシス、プルーン等黒系果実で表現される事も多い種類です。/i,
      ),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("propsのwinesを元にコンポーネントが正しく描画される", () => {
    render(
      <SelectSection
        wines={categoryFavoriteSmallArticleDummyData}
        title="渋い"
        link="sober"
      />,
    );
    expect(
      screen.getByText(categoryFavoriteSmallArticleDummyData[0].title),
    ).toBeInTheDocument();
    expect(
      screen.getByText(categoryFavoriteSmallArticleDummyData[1].title),
    ).toBeInTheDocument();
    expect(
      screen.getByText(categoryFavoriteSmallArticleDummyData[2].title),
    ).toBeInTheDocument();
    expect(
      screen.getByText(categoryFavoriteSmallArticleDummyData[3].title),
    ).toBeInTheDocument();
  });
});
