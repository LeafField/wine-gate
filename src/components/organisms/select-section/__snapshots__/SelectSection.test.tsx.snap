// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectSectionの単体テスト 与えられたtitle及びchildren要素を元にコンポーネントが正しく描画される 1`] = `
<div>
  <section
    class="mt-6 max-w-[1392px] border-b border-gray pb-6"
  >
    <h4
      class="text-heading3"
    >
      渋い
    </h4>
    <p
      class="text-main-text"
    >
      タンニンが強さが強調されていたり、ブラックベリーやカシス、プルーン等黒系果実で表現される事も多い種類です。
      <br />
      タンニンは口の中で渋みを感じさせる成分であり、若いワインでは特に強く感じられることがあります。
      <br />
      カベルネ・ソーヴィニヨンやシラーなどの品種が一般的にこのタイプのワインに含まれます。
      <br />
      主に色が濃く重めの肉料理に合いやすいです。
    </p>
    <div
      class="mt-6 grid grid-cols-1 gap-x-5 gap-y-4 tablet:grid-cols-[repeat(auto-fill,333px)]"
    >
      <a
        aria-label="テストワインその1"
        class="group row-span-3 grid h-full w-full grid-rows-subgrid gap-0 border border-gray transition-colors duration-500 hover:border-blue"
        href="/article/4be1f2ab-b3b3-4599-b1a1-a73d280f5033"
      >
        <div
          class="aspect-video w-full overflow-hidden bg-gray"
        >
          <figure
            class="relative h-full w-full transition-transform duration-300 group-hover:scale-110"
          >
            <img
              alt="ワイン画像"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              sizes="(min-width:768px) 332px,100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
            />
          </figure>
        </div>
        <h3
          class="w-full px-2 text-base text-black"
        >
          テストワインその1
        </h3>
        <div
          class="mt-auto flex flex-row justify-center gap-2 pb-2 pt-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            渋い
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            樽熟成
          </div>
        </div>
      </a>
      <a
        aria-label="テストワインその2"
        class="group row-span-3 grid h-full w-full grid-rows-subgrid gap-0 border border-gray transition-colors duration-500 hover:border-blue"
        href="/article/4be1f2ab-b3b3-4599-b1a1-a73d280f50"
      >
        <div
          class="aspect-video w-full overflow-hidden bg-gray"
        >
          <figure
            class="relative h-full w-full transition-transform duration-300 group-hover:scale-110"
          >
            <img
              alt="ワイン画像"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              sizes="(min-width:768px) 332px,100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
            />
          </figure>
        </div>
        <h3
          class="w-full px-2 text-base text-black"
        >
          テストワインその2
        </h3>
        <div
          class="mt-auto flex flex-row justify-center gap-2 pb-2 pt-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            渋い
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            樽熟成
          </div>
        </div>
      </a>
      <a
        aria-label="テストワインその3"
        class="group row-span-3 grid h-full w-full grid-rows-subgrid gap-0 border border-gray transition-colors duration-500 hover:border-blue"
        href="/article/4be1f2-b3b3-4599-b1a1-a73d280f50"
      >
        <div
          class="aspect-video w-full overflow-hidden bg-gray"
        >
          <figure
            class="relative h-full w-full transition-transform duration-300 group-hover:scale-110"
          >
            <img
              alt="ワイン画像"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              sizes="(min-width:768px) 332px,100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
            />
          </figure>
        </div>
        <h3
          class="w-full px-2 text-base text-black"
        >
          テストワインその3
        </h3>
        <div
          class="mt-auto flex flex-row justify-center gap-2 pb-2 pt-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            渋い
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            樽熟成
          </div>
        </div>
      </a>
      <a
        aria-label="テストワインその4"
        class="group row-span-3 grid h-full w-full grid-rows-subgrid gap-0 border border-gray transition-colors duration-500 hover:border-blue"
        href="/article/4be1f2-b3b3-4599-b1ad1-a73d280f50"
      >
        <div
          class="aspect-video w-full overflow-hidden bg-gray"
        >
          <figure
            class="relative h-full w-full transition-transform duration-300 group-hover:scale-110"
          >
            <img
              alt="ワイン画像"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              sizes="(min-width:768px) 332px,100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
            />
          </figure>
        </div>
        <h3
          class="w-full px-2 text-base text-black"
        >
          テストワインその4
        </h3>
        <div
          class="mt-auto flex flex-row justify-center gap-2 pb-2 pt-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            渋い
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-sm"
          >
            樽熟成
          </div>
        </div>
      </a>
    </div>
    <a
      class="mt-2 flex items-center justify-end text-black transition-colors duration-300 hover:text-blue"
      href="/category/sober"
    >
      <span
        class="text-main-text"
      >
        もっと見る
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrow-right-long h-[w-[clamp(0.875rem,1.54vw,1rem)] ml-[0.125rem] w-[clamp(0.875rem,1.54vw,1rem)]"
        data-icon="arrow-right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M502.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-128-128c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L402.7 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l370.7 0-73.4 73.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l128-128z"
          fill="currentColor"
        />
      </svg>
    </a>
  </section>
</div>
`;
